File: app/src/main/java/io/eodc/planit/adapter/AssignmentViewHolder.java
package io.eodc.planit.adapter;

import android.support.v7.widget.RecyclerView;
import android.view.View;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.TextView;

import butterknife.BindView;
import butterknife.ButterKnife;
import io.eodc.planit.R;
import io.eodc.planit.db.Assignment;

/**
 * Holder for information and attributes of the assignment view
 *
 * @author 2n
 */
public class AssignmentViewHolder extends RecyclerView.ViewHolder {
    boolean isExpanded = false;
    Assignment assignment;

    @BindView(R.id.header)              LinearLayout    layoutHeader;
    @BindView(R.id.ic_due_date)         ImageView       iconDueDate;
    @BindView(R.id.expand_button)       ImageView       iconExpand;
    @BindView(R.id.class_color)         ImageView       imageClassColor;
    @BindView(R.id.header_label)        TextView        textHeader;
    @BindView(R.id.assignment_name)     TextView        textAssignmentName;
    @BindView(R.id.class_and_type)      TextView        textClassType;
    @BindView(R.id.due_date)            TextView        textDueDate;
    @BindView(R.id.notes_text)          TextView        textNotes;

    /**
     * Constructs an instance of an AssignmentViewHolder
     *
     * @param itemView The view to bind to this holder
     */
    AssignmentViewHolder(View itemView) {
        super(itemView);
        ButterKnife.bind(this, itemView);
    }

    public Assignment getAssignment() {
        return assignment;
    }

    void hideDueDate() {
        textDueDate.setVisibility(View.GONE);
        iconDueDate.setVisibility(View.GONE);
    }

    void showDueDate() {
        textDueDate.setVisibility(View.VISIBLE);
        iconDueDate.setVisibility(View.VISIBLE);
    }

    /**
     * Hides the textNotes section of the view
     */
    void shrinkNotes() {
        iconExpand.animate()
                .rotation(0f);
        textNotes.setVisibility(View.GONE);
        isExpanded = false;
    }

    /**
     * Shows the textNotes section of the vie
     */
    void expandNotes() {
        iconExpand.animate()
                .rotation(-180f);
        textNotes.setVisibility(View.VISIBLE);
        isExpanded = true;
    }

    void handleNoteClick() {
        if (isExpanded) {
            shrinkNotes();
        } else {
            expandNotes();
        }
    }
}

File: app/src/main/java/io/eodc/planit/adapter/AssignmentViewHolder.kt
package io.eodc.planit.adapter

import android.content.Context
import android.os.Build
import android.os.VibrationEffect
import android.os.Vibrator
import android.view.View
import android.widget.ImageView
import android.widget.LinearLayout
import android.widget.TextView
import androidx.appcompat.app.AppCompatActivity
import com.ernestoyaquello.dragdropswiperecyclerview.DragDropSwipeAdapter
import io.eodc.planit.db.Assignment
import io.eodc.planit.fragment.ModifyAssignmentFragment
import kotlinx.android.synthetic.main.item_assignment.view.*

/**
 * Holder for information and attributes of the assignment view
 *
 * @author 2n
 */
class AssignmentViewHolder
/**
 * Constructs an instance of an AssignmentViewHolder
 *
 * @param itemView The view to bind to this holder
 */
internal constructor(itemView: View) : DragDropSwipeAdapter.ViewHolder(itemView) {
    private var isExpanded = false
    var assignment: Assignment? = null

    internal var layoutHeader: LinearLayout = itemView.header
    internal var iconExpand: ImageView = itemView.btn_expand
    internal var imageClassColor: ImageView = itemView.border_class_color
    internal var textHeader: TextView = itemView.text_header
    internal var textAssignmentName: TextView = itemView.textHeaderTitle
    internal var textClassType: TextView = itemView.text_class_type
    internal var textDueDate: TextView = itemView.text_due
    internal var textNotes: TextView = itemView.text_notes

    internal fun hideDueDate() {
        textDueDate.visibility = View.GONE
    }

    internal fun showDueDate() {
        textDueDate.visibility = View.VISIBLE
    }

    /**
     * Hides the textNotes section of the view
     */
    private fun shrinkNotes() {
        iconExpand.animate()
                .rotation(0f)
        textNotes.visibility = View.GONE
        isExpanded = false
    }

    fun editAssignment(activity: AppCompatActivity) {
        val editFragment = ModifyAssignmentFragment.newInstance(assignment!!)

        editFragment.show(activity.supportFragmentManager, null)

        val v = activity.getSystemService(Context.VIBRATOR_SERVICE) as Vibrator
        @Suppress("DEPRECATION")
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O)
            v.vibrate(VibrationEffect
                    .createOneShot(10, VibrationEffect.DEFAULT_AMPLITUDE))
        else
            v.vibrate(10)
    }

    /**
     * Shows the textNotes section of the vie
     */
    private fun expandNotes() {
        iconExpand.animate()
                .rotation(-180f)
        textNotes.visibility = View.VISIBLE
        isExpanded = true
    }

    internal fun handleNoteClick() {
        if (isExpanded) {
            shrinkNotes()
        } else {
            expandNotes()
        }
    }
}
